plugins/org.eclipse.n4js/src/org/eclipse/n4js/typesystem/N4JSTypeSystem.java

```diff
@@ -321,4 +326,51 @@ public TypeRef createSimplifiedIntersection(List<TypeRef> typeRefs, Resource res
 			return null;
 		}
 	}
+
+	/** @return the return {@link TypeRef} for a given {@link FunctionOrFieldAccessor} */
+	public TypeRef getReturnTypeRef(FunctionOrFieldAccessor fofa) {
+		TypeRef inferredType = tau(fofa);
+		if (inferredType instanceof FunctionTypeExprOrRef) {
+			return getActualReturnTypeRef(fofa, (FunctionTypeExprOrRef) inferredType);
+		}
+		if (inferredType instanceof ParameterizedTypeRef) {
+			return inferredType;
+		}
+		return null;
+	}
+
+	private TypeRef getActualReturnTypeRef(FunctionOrFieldAccessor fofa, FunctionTypeExprOrRef inferredType) {
+		if (inferredType == null) {
+			return null;
+		}
+		TypeRef typeRef = inferredType.getReturnTypeRef();
+		if (fofa instanceof FunctionDefinition) {
+			FunctionDefinition fDef = (FunctionDefinition) fofa;
+			Type tFun = fDef.getDefinedType();
+			if (fofa.isAsync()) {
+				typeRef = getTypeArgumentOfReturnType(fDef, tFun, 0);
+			}
+			if (fDef.isGenerator()) {
+				typeRef = getTypeArgumentOfReturnType(fDef, tFun, 1);
```

Comment:
don't we already have a utility/helper method for this, somewhere? ("this" including special handling of generators and async) If not, please create one in one of the utility/helper classes (e.g. TypeUtils) in a reusable form.